File: Trabalho_3/src/com/company/Conjunto.java
/**
 * @file   Conjunto.java
 * @brief  Arquivo dedicado à classe abstrata que.
 * @author Gustavo Rangel Bandeira Cauduro
 * @date   2022-01-15
 */

package com.company;

public abstract class Conjunto<T> {
    public abstract Boolean contemElemento(T elemento);
    Conjunto<T> adicionarElemento(T elemento) {
        return new ConjuntoElemento<>(elemento, this);
    }
    Conjunto<T> uniao(Conjunto<T> conjunto ) {
        return new Uniao<>(this, conjunto);
    }
    Conjunto<T> intersecao(Conjunto<T> conjunto) {
        return new Intersecao<>(this, conjunto);
    }
    Conjunto<T> diferenca(Conjunto<T> conjunto) {
        return new Diferenca<>(this, conjunto);
    }
    Conjunto<T> complemento() {
        return new Complemento<>(this);
    }
}

File: Trabalho_3_C#/Trabalho_3_CSharp/Conjunto.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Trabalho_3_CSharp
{
    public abstract class Conjunto<T> {

        public abstract Boolean contemElemento(T elemento);
        public Conjunto<T> adicionarElemento(T elemento) {
            return new ConjuntoElemento<T>(elemento, this);
        }
        public Conjunto<T> uniao(Conjunto<T> conjunto) {
            return new Uniao<T>(this, conjunto);
        }
        public Conjunto<T> intersecao(Conjunto<T> conjunto) {
            return new Intersecao<T>(this, conjunto);
        }
        public Conjunto<T> diferenca(Conjunto<T> conjunto) {
            return new Diferenca<T>(this, conjunto);
        }
        public Conjunto<T> complemento() {
            return new Complemento<T>(this);
        }
    }
}
